Fixed #30095 -- Fixed system check for RangeField/ArrayField.choices with lists and tuples.

hramezani opened the pull request:
[ticket 30095](https://code.djangoproject.com/ticket/30095)

charettes commented:
Any thoughts about completely overriding `RangeField._check_choices` instead?

I assume `ArrayField` suffers from similar problems.

hramezani commented:
@charettes overriding `RangeField._check_choices` require a lot of code duplication. I thought it is easier to override `is_value()` based on @carltongibson [comment](https://code.djangoproject.com/ticket/30095#comment:6) on the ticket.

> Maybe raising the is_value() helper to class level would allow avoiding too much duplication

felixxm commented on django/db/models/fields/__init__.py:
```diff
@@ -237,14 +237,15 @@ def _check_field_name(self):
-        def is_value(value, accept_promise=True):
-            return isinstance(value, (str, Promise) if accept_promise else str) or not is_iterable(value)
-
-        if is_value(self.choices, accept_promise=False):
```
We use `accept_promise=False` only here it's misleading to me, I opened #12000 to fix this.

felixxm commented on tests/postgres_tests/test_ranges.py:
```diff
@@ -88,6 +89,16 @@ def test_model_set_on_base_field(self):
+                    ((50, 100), '51-100'),
+                )
+            )
+        self.assertEqual(Model._meta.get_field('field').check(), [])
```
We should also add test for grouped choices.

carltongibson commented:
(Note: My suggestion on the ticket was a __maybe__ we reduce duplication...)

pope1ni commented on django/db/models/fields/__init__.py:
```diff
@@ -237,14 +237,15 @@ def _check_field_name(self):
         else:
             return []
 
+    @classmethod
```
If you're using `@classmethod` because `self` wasn't used, then you may as well use `@staticmethod` as `cls` won't be used either...

hramezani replied:
Good point, Thanks.

hramezani commented:
@felixxm 
I moved `is_value()` to class level and renamed it to  `_choices_is_value()` and Added a grouped choice to the test choices.
I will prepare another patch for `ArrayField` we successfully merged this one.

hramezani commented:
@felixxm Do we need something else on this patch?

felixxm commented:
@hramezani I think we should add a fix for `ArrayField` to this PR.

hramezani commented:
@felixxm I added a commit to fix `ArrayField` with lists in choices.

felixxm commented:
@hramezani Thanks for updates :+1: I simplified tests, changed `_choices_is_value()` to `classmethods`'s, used `super()` in them and added lists and tuples to `ArrayField`'s and `RangeField`'s checks, because both support them.

hramezani commented:
> @hramezani Thanks for updates  I simplified tests, changed `_choices_is_value()` to `classmethods`'s, used `super()` in them and added lists and tuples to `ArrayField`'s and `RangeField`'s checks, because both support them.

Thanks @felixxm. Good as always :+1:
